File: Avail/src/com/avail/interpreter/levelTwo/L2Chunk.java
/**
 * L2Chunk.java
 * Copyright © 1993-2013, Mark van Gulik and Todd L Smith.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.interpreter.levelTwo;

import java.lang.ref.*;
import java.util.*;
import java.util.concurrent.locks.ReentrantLock;
import com.avail.annotations.*;
import com.avail.descriptor.*;
import com.avail.optimizer.L2Translator;
import com.avail.interpreter.levelTwo.register.*;

/**
 * A Level Two chunk represents an optimized implementation of a {@linkplain
 * CompiledCodeDescriptor compiled code object}.
 *
 * <p>
 * The chunks are held onto by weak references in a list (keyed by chunk index).
 * When a chunk expires due to not being referred to by any code or
 * continuations, its weak reference is added to a queue from which chunk index
 * recycling takes place.  The weak references also keep track of the contingent
 * methods.  The methods maintain the reverse relation
 * by keeping track of the indices of all chunks that depend on them.  When an
 * method changes (due to a method being added or removed), the
 * dependent chunks can be marked as invalid and eviscerated (to reclaim
 * memory).  When an attempt is made to use an invalidated chunk by invoking a
 * compiled code object or returning into a continuation, the reference to the
 * chunk is replaced by a reference to the default chunk (and a continuation
 * gets its offset set to the level one dispatch loop).  When all references to
 * the chunk have been so replaced, the chunk's weak reference will appear on
 * the ReferenceQueue, allowing the index to be recycled.
 * </p>
 *
 * <p>
 * Eventually we can limit the number of valid chunks by linking the weak
 * references together into an LRU ring.  When adding a new chunk to a "full"
 * ring, the oldest element can simply be invalidated, removing it from the
 * ring.  The level two interpreter is already instrumented to call
 * moveToHead() at appropriate times.
 * </p>
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 */
public final class L2Chunk
{
	/**
	 * The unique integer that identifies this chunk. Weak references are
	 * used to determine when it is safe to recycle an index for a new
	 * chunk.
	 */
	int index;

	/**
	 * The number of {@linkplain L2ObjectRegister object registers} that
	 * this chunk uses (including the fixed registers).  Having the number of
	 * needed object registers stored separately allows the register list to be
	 * dynamically expanded as needed only when starting or resuming a
	 * {@link ContinuationDescriptor continuation}.
	 */
	int numObjects;

	/**
	 * The number of {@linkplain L2IntegerRegister integer registers} that
	 * are used by this chunk. Having this recorded separately allows the
	 * register list to be dynamically expanded as needed only when starting
	 * or resuming a continuation.
	 */
	int numIntegers;

	/**
	 * The number of {@linkplain L2FloatRegister floating point registers}
	 * that are used by this chunk. Having this recorded separately allows
	 * the register list to be dynamically expanded as needed only when
	 * starting or resuming a continuation.
	 */
	int numDoubles;

	/**
	 * A flag indicating whether this chunk has been reached by the garbage
	 * collector in the current scavenge cycle. If it's still clear at flip
	 * time, the chunk is unreferenced and can be reclaimed.
	 *
	 * TODO: [MvG] This is not used by the current (2011.05.11)
	 * Avail-on-Java VM.
	 */
	boolean saved;

	/**
	 * A flag indicating whether this chunk is valid or if it has been
	 * invalidated by the addition or removal of a method signature.
	 */
	boolean valid;

	/**
	 * The sequence of {@link L2Instruction}s that make up this L2Chunk.
	 */
	@SuppressWarnings("null")
	public L2Instruction [] instructions;

	/**
	 * Answer the Avail {@linkplain PojoDescriptor pojo} associated with this
	 * L2Chunk.
	 */
	public final AvailObject chunkPojo = PojoDescriptor.newPojo(
			RawPojoDescriptor.identityWrap(this),
			PojoTypeDescriptor.forClass(this.getClass()))
		.makeShared();

	/**
	 * Answer an integer that uniquely identifies this chunk.  This allows a
	 * chunk to be invalidated given only its index and {@link
	 * #allChunksWeakly}.  In fact, the L2Chunk itself may already have been
	 * garbage collected by that point!
	 *
	 * @return The chunk's identifying index.
	 */
	public int index ()
	{
		return index;
	}

	/**
	 * Answer the number of floating point registers used by this chunk.
	 *
	 * @return The count of float registers.
	 */
	public int numDoubles ()
	{
		return numDoubles;
	}

	/**
	 * Answer the number of integer registers used by this chunk.
	 *
	 * @return The count of integer registers.
	 */
	public int numIntegers ()
	{
		return numIntegers;
	}

	/**
	 * Answer the number of object registers used by this chunk.
	 *
	 * @return The count of object registers.
	 */
	public int numObjects ()
	{
		return numObjects;
	}

	/**
	 * Answer whether this chunk is still valid.  A {@linkplain
	 * ContinuationDescriptor continuation} or {@linkplain
	 * CompiledCodeDescriptor raw function} may refer to an invalid chunk, but
	 * attempts to resume or invoke (respectively) such a chunk are detected and
	 * cause the {@link #unoptimizedChunk()} to be substituted instead.  We
	 * don't have to worry about an Interpreter holding onto a chunk when it
	 * becomes invalid, because invalidation can only happen when the runtime
	 * temporarily inhibits running Avail code, and all fibers have had their
	 * continuation reified to a level-one-coherent state.
	 *
	 * @return Whether this chunk is still valid.
	 */
	public boolean isValid ()
	{
		return valid;
	}

	@Override
	public String toString ()
	{
		final StringBuilder builder = new StringBuilder();
		if (index() == 0)
		{
			return "Default chunk #0";
		}
		builder.append("Chunk #");
		builder.append(index());
		builder.append("\n");
		if (!isValid())
		{
			builder.append("\t(INVALID)\n");
		}
		final L2InstructionDescriber describer =
			new L2InstructionDescriber();
		int offset = 0;
		for (final L2Instruction instruction : instructions)
		{
			builder.append(String.format("\t#%-3d ", offset));
			final StringBuilder tempStream = new StringBuilder(100);
			describer.describe(instruction, this, tempStream);
			builder.append(
				tempStream.toString().replace("\n", "\n\t\t"));
			builder.append("\n");
			offset++;
		}
		return builder.toString();
	}

	/**
	 * A {@link WeakChunkReference} is the mechanism by which {@linkplain
	 * L2Chunk level two chunks} are recycled by the Java garbage collector.
	 * When a chunk is only weakly reachable (i.e., there are no strong or soft
	 * references to it), the index reserved for that chunk becomes eligible for
	 * recycling.
	 *
	 * @author Mark van Gulik &lt;mark@availlang.org&gt;
	 */
	static class WeakChunkReference
	extends WeakReference<L2Chunk>
	{
		/**
		 * The finalization queue onto which {@linkplain L2Chunk level two
		 * chunks}' {@linkplain WeakChunkReference weak references} will be
		 * placed upon expiration. There is no special process to remove them
		 * from here. Rather, an element of this queue is consumed when needed
		 * for {@linkplain L2Chunk#allocate(A_RawFunction, int, int, int, List,
		 * Set) allocation} of a new chunk. If this queue is empty, a fresh
		 * index is allocated.
		 */
		static final ReferenceQueue<L2Chunk> recyclingQueue =
			new ReferenceQueue<>();

		/**
		 * The {@linkplain L2Chunk#index} of the {@linkplain L2Chunk level two
		 * chunk} to which this reference either refers or once referred.
		 */
		final int index;

		/**
		 * The list of {@linkplain MethodDescriptor methods} on which the
		 * referent chunk depends. If one of these methods changes (due to
		 * adding or removing a {@linkplain DefinitionDescriptor method
		 * implementation}), this chunk will be immediately invalidated.
		 */
		final Set<A_Method> contingentMethods;

		/**
		 * Construct a new {@link WeakChunkReference}.
		 *
		 * @param chunk
		 *        The chunk to be wrapped with a weak reference.
		 * @param contingentMethods
		 *        The {@link Set} of {@linkplain MethodDescriptor methods} on
		 *        which this chunk depends.
		 */
		public WeakChunkReference (
			final L2Chunk chunk,
			final Set<A_Method> contingentMethods)
		{
			super(chunk, recyclingQueue);
			this.index = chunk.index();
			this.contingentMethods = contingentMethods;
		}
	}

	/**
	 * A list of {@linkplain WeakChunkReference weak chunk references} to every
	 * {@linkplain L2Chunk level two chunk}. The chunks are wrapped within a
	 * {@link WeakChunkReference} and placed in the list at their {@linkplain
	 * L2Chunk#index}, which the weak chunk reference also records. Some of the
	 * weak references may have a null {@linkplain WeakReference#get()
	 * referent}, indicating the chunk at that position has either been
	 * reclaimed or will appear on the {@link WeakChunkReference#recyclingQueue
	 * finalization queue} shortly.
	 */
	private static final List<WeakChunkReference> allChunksWeakly =
		new ArrayList<>(100);

	/**
	 * The level two wordcode offset to which to jump when returning into a
	 * continuation that's running the {@linkplain #unoptimizedChunk unoptimized
	 * chunk}.
	 *
	 * @return A level two wordcode offset.
	 */
	public static int offsetToContinueUnoptimizedChunk ()
	{
		// This is hard-coded, but cross-checked by
		// L2Translator#createChunkForFirstInvocation().
		return 7;
	}

	/**
	 * Return the number of times to invoke a {@linkplain CompiledCodeDescriptor
	 * compiled code} object, <em>after an invalidation</em>, before attempting
	 * to optimize it again.
	 *
	 * @return The number of invocations before post-invalidate reoptimization.
	 */
	public static int countdownForInvalidatedCode ()
	{
		return 10;
	}

	/**
	 * Return the number of times to invoke a {@linkplain CompiledCodeDescriptor
	 * compiled code} object, <em>after creation</em>, before attempting to
	 * optimize it for the first time.
	 *
	 * @return The number of invocations before initial optimization.
	 */
	public static int countdownForNewCode ()
	{
		return 10;
	}

	/**
	 * Return the number of times to invoke a {@linkplain CompiledCodeDescriptor
	 * compiled code} object, <em>after optimization</em>, before attempting to
	 * optimize it again with more effort.
	 *
	 * @return The number of invocations before attempting to improve the
	 *         optimization.
	 */
	public static int countdownForNewlyOptimizedCode ()
	{
		// TODO: [MvG] Set this to something sensible when optimization levels
		// are implemented.
		return 1000000000;
	}

	/**
	 * The {@linkplain ReentrantLock lock} that guards access to the table of
	 * {@linkplain L2Chunk chunks}.
	 */
	private static final ReentrantLock chunksLock = new ReentrantLock();

	/**
	 * Allocate and set up a new {@linkplain L2Chunk level two chunk} with the
	 * given information. If {@code code} is non-null, set it up to use the new
	 * chunk for subsequent invocations.
	 *
	 * @param code
	 *        The {@linkplain CompiledCodeDescriptor code} for which to use the
	 *        new level two chunk, or null for the initial unoptimized chunk.
	 * @param numObjects
	 *        The number of {@linkplain L2ObjectRegister object registers} that
	 *        this chunk will require.
	 * @param numIntegers
	 *        The number of {@linkplain L2IntegerRegister integer registers}
	 *        that this chunk will require.
	 * @param numFloats
	 *        The number of {@linkplain L2FloatRegister floating point
	 *        registers} that this chunk will require.
	 * @param theInstructions
	 *        A {@link List} of {@link L2Instruction}s that prescribe what to do
	 *        in place of the level one nybblecodes.
	 * @param contingentMethods
	 *        A {@link Set} of {@linkplain MethodDescriptor methods} on which
	 *        the level two chunk depends.
	 * @return The new level two chunk.
	 */
	public static L2Chunk allocate (
		final @Nullable A_RawFunction code,
		final int numObjects,
		final int numIntegers,
		final int numFloats,
		final List<L2Instruction> theInstructions,
		final Set<A_Method> contingentMethods)
	{
		final L2Chunk chunk = create(
			numObjects,
			numIntegers,
			numFloats,
			theInstructions);
		final int index;
		chunksLock.lock();
		try
		{
			final ReferenceQueue<L2Chunk> queue =
				WeakChunkReference.recyclingQueue;
			final Reference<? extends L2Chunk> recycledReference =
				queue.poll();
			if (recycledReference != null)
			{
				// Recycle the reference. Nobody referred to the chunk, so it
				// has already been garbage collected and nulled from its weak
				// reference. It may or may not have been invalidated already,
				// so clean it up if necessary.
				final WeakChunkReference oldReference =
					(WeakChunkReference) recycledReference;
				for (final A_Method method : oldReference.contingentMethods)
				{
					method.removeDependentChunkIndex(oldReference.index);
				}
				oldReference.contingentMethods.clear();
				index = oldReference.index;
			}
			else
			{
				// Nothing available for recycling. Make room for it at the end.
				index = allChunksWeakly.size();
				allChunksWeakly.add(null);
			}
			chunk.index = index;
			final WeakChunkReference newReference = new WeakChunkReference(
				chunk,
				contingentMethods);
			allChunksWeakly.set(index, newReference);
		}
		finally
		{
			chunksLock.unlock();
		}

		// Now that the index has been assigned, connect the dependency. Since
		// connecting the dependency may grow some sets, make sure the Avail GC
		// (not yet implemented in Java) can be invoked safely. To assist this,
		// make sure the code is referring to the chunk being set up, to avoid
		// having it garbage collected before we have a chance to install it.
		if (code != null)
		{
			code.setStartingChunkAndReoptimizationCountdown(
				chunk,
				L2Chunk.countdownForNewlyOptimizedCode());
		}
		for (final A_Method method : contingentMethods)
		{
			method.addDependentChunkIndex(index);
		}

		return chunk;
	}

	/**
	 * A method has changed. This means a method definition (or a forward or an
	 * abstract declaration) has been added or removed from the method, and the
	 * specified chunk previously expressed an interest in change notifications.
	 * This must have been because it was optimized in a way that relied on some
	 * aspect of the available definitions (e.g., monomorphic inlining), so we
	 * need to invalidate the chunk now, so that an attempt to invoke it or
	 * return into it will be detected and converted into using the {@linkplain
	 * #unoptimizedChunk unoptimized chunk}. Also remove this chunk's index from
	 * all methods on which it was depending.  Do not add the chunk's reference
	 * to the reference queue, since it may still be referenced by code or
	 * continuations that need to detect that it is now invalid.
	 *
	 * @param chunkIndex The index of the chunk to invalidate.
	 */
	public static void invalidateChunkAtIndex (final int chunkIndex)
	{
		chunksLock.lock();
		try
		{
			final WeakChunkReference ref = allChunksWeakly.get(chunkIndex);
			assert ref.index == chunkIndex;
			final L2Chunk chunk = ref.get();
			if (chunk != null)
			{
				chunk.valid = false;
				// The empty tuple is already shared, so we don't need to share
				// it here.
				chunk.instructions = new L2Instruction[0];
			}
			final Set<A_Method> methods = ref.contingentMethods;
			for (final A_Method method : methods)
			{
				method.removeDependentChunkIndex(chunkIndex);
			}
			ref.contingentMethods.clear();
		}
		finally
		{
			chunksLock.unlock();
		}
	}

	/**
	 * Create a new {@linkplain L2Chunk level two chunk} with the given
	 * information.
	 *
	 * @param numObjects
	 * @param numIntegers
	 * @param numFloats
	 * @param theInstructions
	 * @return
	 */
	private static L2Chunk create (
		final int numObjects,
		final int numIntegers,
		final int numFloats,
		final List<L2Instruction> theInstructions)
	{
		final L2Chunk chunk = new L2Chunk();
		// A new chunk starts out saved and valid.
		chunk.saved = true;
		chunk.valid = true;
		chunk.numObjects = numObjects;
		chunk.numIntegers = numIntegers;
		chunk.numDoubles = numFloats;
		chunk.instructions =
			theInstructions.toArray(new L2Instruction[theInstructions.size()]);
		return chunk;
	}

	/**
	 * The special {@linkplain L2Chunk level two chunk} that is used to
	 * interpret level one nybblecodes until a piece of {@linkplain
	 * CompiledCodeDescriptor compiled code} has been executed some number of
	 * times.
	 */
	private static final L2Chunk unoptimizedChunk =
		L2Translator.createChunkForFirstInvocation();

	static
	{
		assert unoptimizedChunk.index() == 0;
		assert allChunksWeakly.size() == 1;
	}

	/**
	 * Return the special {@linkplain L2Chunk level two chunk} that is used to
	 * interpret level one nybblecodes until a piece of {@linkplain
	 * CompiledCodeDescriptor compiled code} has been executed some threshold
	 * number of times.
	 *
	 * @return The special {@linkplain #unoptimizedChunk unoptimized chunk}.
	 */
	public static L2Chunk unoptimizedChunk ()
	{
		return unoptimizedChunk;
	}
}

File: avail/src/main/kotlin/avail/interpreter/levelTwo/L2Chunk.kt
/*
 * L2Chunk.kt
 * Copyright © 1993-2022, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package avail.interpreter.levelTwo

import avail.AvailRuntime
import avail.AvailRuntimeSupport
import avail.descriptor.fiber.FiberDescriptor
import avail.descriptor.functions.A_Continuation
import avail.descriptor.functions.A_RawFunction
import avail.descriptor.functions.A_RawFunction.Companion.methodName
import avail.descriptor.functions.A_RawFunction.Companion.setStartingChunkAndReoptimizationCountdown
import avail.descriptor.functions.CompiledCodeDescriptor
import avail.descriptor.methods.A_ChunkDependable
import avail.descriptor.methods.MethodDescriptor
import avail.descriptor.pojos.PojoDescriptor
import avail.descriptor.pojos.RawPojoDescriptor.Companion.identityPojo
import avail.descriptor.representation.AvailObject
import avail.descriptor.sets.A_Set
import avail.descriptor.sets.SetDescriptor.Companion.emptySet
import avail.interpreter.execution.Interpreter
import avail.interpreter.execution.Interpreter.Companion.log
import avail.interpreter.levelTwo.L2Chunk.Generation
import avail.interpreter.levelTwo.L2Chunk.InvalidationReason.EVICTION
import avail.interpreter.levelTwo.L2JVMChunk.Companion.unoptimizedChunk
import avail.optimizer.ExecutableChunk
import avail.optimizer.jvm.JVMChunk
import avail.optimizer.jvm.JVMTranslator
import avail.optimizer.jvm.ReferencedInGeneratedCode
import avail.performance.Statistic
import avail.performance.StatisticReport.L2_OPTIMIZATION_TIME
import avail.utility.safeWrite
import java.lang.ref.WeakReference
import java.util.ArrayDeque
import java.util.Collections.synchronizedSet
import java.util.Deque
import java.util.concurrent.locks.ReentrantLock
import java.util.concurrent.locks.ReentrantReadWriteLock
import java.util.logging.Level
import javax.annotation.concurrent.GuardedBy
import kotlin.concurrent.withLock

/**
 * A Level Two chunk represents an optimized implementation of a
 * [compiled&#32;code&#32;object][CompiledCodeDescriptor].
 *
 * An [A_RawFunction] refers to the L2Chunk that it should run in its place.  An
 * [A_Continuation] also refers to the L2Chunk that allows the continuation to
 * be returned into, restarted, or resumed after an interrupt. The [Generation]
 * mechanism maintains approximate age information of chunks, in particular how
 * long it has been since a chunk was last used, so that the least recently used
 * chunks can be evicted when there are too many chunks in memory.
 *
 * A chunk also keeps track of the methods that it depends on, and the methods
 * keep track of which chunks depend on them.  New method definitions can be
 * added – or existing ones removed – only while all fiber execution is paused.
 * At this time, the chunks that depend on the changed method are marked as
 * invalid.  Each [A_RawFunction] associated (1:1) with an invalidated chunk has
 * its [A_RawFunction.startingChunk] reset to the default chunk.  Existing
 * continuations may still be referring to the invalid chunk – but not Java call
 * frames, since all fibers are paused.  When resuming a continuation, its
 * chunk's validity is immediately checked, and if it's invalid, the default
 * chunk is resumed at a suitable entry point instead.
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 *
 * @property code
 *   The code that was translated to L2.  Null for the default (L1) chunk.
 * @property offsetAfterInitialTryPrimitive
 *   The level two offset at which to start if the corresponding [A_RawFunction]
 *   is a primitive, and it has already been attempted and failed.  If it's not
 *   a primitive, this is the offset of the start of the code (0).
 * @property contingentValues
 *   The set of [contingent&#32;values][A_ChunkDependable] on which this chunk
 *   depends. If one of these changes significantly, this chunk must be
 *   invalidated (at which time this set will be emptied).
 *
 * @constructor
 * Create a new [L2Chunk] with the given information.
 */
abstract class L2Chunk protected constructor(
	val code: A_RawFunction?,
	val offsetAfterInitialTryPrimitive: Int,
	private var contingentValues: A_Set)
{
	/** The [ExecutableChunk] to run. */
	abstract val executableChunk: ExecutableChunk

	/** The [L2AbstractInstruction]s constituting this chunk. */
	abstract val instructions: List<L2AbstractInstruction>

	/** The [WeakReference] that points to this [L2Chunk]. */
	@Suppress("LeakingThis")
	val weakReference = WeakReference(this)

	/**
	 * An indication of how recently this chunk has been accessed, expressed as
	 * a reference to a [Generation].
	 */
	@Volatile
	var generation: Generation? = Generation.newest

	/**
	 * A group of chunks with approximately equal most-recent access time.
	 */
	class Generation
	{
		/**
		 * The weak set of [L2Chunk]s in this generation.
		 */
		private val chunks = synchronizedSet(HashSet<WeakReference<L2Chunk>>())

		override fun toString(): String
		{
			return super.toString() + " (size=" + chunks.size + ")"
		}

		companion object
		{
			/**
			 * The [Deque] of [Generation]s.  New ones are added with
			 * [Deque.addFirst], and older ones are removed with
			 * [Deque.removeLast] (while invalidating the contained chunks).
			 */
			@GuardedBy("generationsLock")
			private val generations: Deque<Generation> = ArrayDeque()

			/** The lock for accessing the [Deque] of [Generation]s. */
			private val generationsLock = ReentrantReadWriteLock()

			/**
			 * A [Generation] that has not yet been added to the [generations]
			 * [Deque].  When this becomes fuller than approximately
			 * [maximumNewestGenerationSize], queue it and create a new one.
			 */
			@Volatile
			var newest = Generation()

			/**
			 * The maximum number of chunks to place in this generation before
			 * creating a newer one.  If the working set of chunks is larger
			 * than this, there is a risk of thrashing (invalidating and
			 * recompiling a lot of [L2Chunk]s), which is balanced against
			 * over-consumption of memory by chunks.
			 */
			private const val maximumNewestGenerationSize = 1_000

			/**
			 * The approximate maximum number of chunks that should exist at any
			 * time.  When there are significantly more chunks than this, the
			 * ones in the oldest generations will be invalidated.
			 */
			private const val maximumTotalChunkCount = 10_000

			/**
			 * Record a newly created chunk in the latest generation, triggering
			 * eviction of some of the least recently used chunks if necessary.
			 *
			 * @param newChunk
			 *   The new chunk to track.
			 */
			fun addNewChunk(newChunk: L2Chunk)
			{
				newChunk.generation = newest
				newest.chunks.add(newChunk.weakReference)
				if (newest.chunks.size > maximumNewestGenerationSize)
				{
					generationsLock.safeWrite {
						var lastGenerationToKeep = newest
						generations.addFirst(newest)
						newest = Generation()
						var liveCount = 0
						for (gen in generations)
						{
							val genSize = gen.chunks.size
							liveCount += genSize
							if (liveCount >= maximumTotalChunkCount) break
							lastGenerationToKeep = gen
						}
						// Remove the obsolete generations, gathering the chunks.
						val chunksToInvalidate =
							mutableListOf<WeakReference<L2Chunk>>()
						while (generations.last !== lastGenerationToKeep)
						{
							chunksToInvalidate.addAll(
								generations.removeLast().chunks)
						}
						// Remove empty generations that would otherwise be kept.
						val toKeep =
							generations.filter { it.chunks.isNotEmpty() }
						generations.clear()
						generations.addAll(toKeep)
						if (chunksToInvalidate.isNotEmpty())
						{
							// Queue a task to safely invalidate the evicted
							// chunks.
							AvailRuntime.currentRuntime().whenSafePointDo(
								FiberDescriptor.bulkL2InvalidationPriority)
							{
								invalidationLock.withLock {
									chunksToInvalidate.forEach {
										it.get()?.invalidate(EVICTION)
									}
								}
							}
						}
					}
				}
			}

			/**
			 * Deal with the fact that the given chunk has just been invoked,
			 * resumed, restarted, or otherwise continued.  Optimize for the
			 * most common case that the chunk is already in the newest
			 * generation, but also make it reasonably quick to move it there
			 * from an older generation.
			 *
			 * @param chunk
			 *   The [L2Chunk] that has just been used.
			 */
			fun usedChunk(chunk: L2Chunk)
			{
				val theNewest = newest
				val oldGen = chunk.generation
				if (oldGen === theNewest)
				{
					// The chunk is already in the newest generation, which
					// should be the most common case by far.  Do nothing.
					return
				}
				// Move the chunk to the newest generation.  Create a newer
				// generation if it fills up.
				oldGen?.run { chunks.remove(chunk.weakReference) }
				theNewest.chunks.add(chunk.weakReference)
				chunk.generation = theNewest
				if (theNewest.chunks.size > maximumNewestGenerationSize)
				{
					generationsLock.safeWrite {
						// Now that we have the lock, make sure a new generation
						// hasn't been introduced by some other thread.
						if (theNewest == newest)
						{
							generations.add(newest)
							newest = Generation()
							// Even though simply using a chunk doesn't exert
							// any cache pressure, we might accumulate a bunch
							// of empty generations that simply take up space.
							// Be generous and only bother scanning if there are
							// so many generations that there's definitely at
							// least one empty one.
							if (generations.size > maximumTotalChunkCount)
							{
								generations.removeIf { it.chunks.isEmpty() }
							}
						}
					}
				}
			}

			/**
			 * An [L2Chunk] has been invalidated. Remove it from its generation.
			 *
			 * @param chunk
			 *   The invalidated [L2Chunk] to remove from its generation.
			 */
			fun removeInvalidatedChunk(chunk: L2Chunk)
			{
				chunk.generation?.let {
					it.chunks.remove(chunk.weakReference)
					chunk.generation = null
				}
			}
		}
	}

	/**
	 * A flag indicating whether this chunk is valid or if it has been
	 * invalidated by the addition or removal of a method signature.  It doesn't
	 * have to be `volatile`, since it can only be set when Avail code
	 * execution is temporarily suspended in all fibers, which involves
	 * synchronization (and therefore memory coherence) before it can start
	 * running again.
	 */
	@get:ReferencedInGeneratedCode
	var isValid = true
		private set

	/**
	 * Answer the Avail [pojo][PojoDescriptor] associated with this L2Chunk.
	 */
	@Suppress("LeakingThis")
	val chunkPojo: AvailObject = identityPojo(this).makeShared()

	fun name(): String = name(code)

	override fun toString(): String
	{
		if (this == unoptimizedChunk)
		{
			return "Default chunk"
		}
		val builder = StringBuilder()
		if (!isValid)
		{
			builder.append("[INVALID] ")
		}
		builder.append(String.format(
			"Chunk #%08x",
			System.identityHashCode(this)))
		code?.run {
			builder.append(" for $methodName")
		}
		return builder.toString()
	}

	/**
	 * Called just before running the [JVMChunk] inside this [L2Chunk].
	 * This gives the opportunity for logging the chunk execution.
	 *
	 * @param offset
	 *   The L2 offset at which we're about to start or continue running this
	 *   chunk.
	 */
	fun beforeRunChunk(offset: Int)
	{
		if (Interpreter.debugL2)
		{
			log(
				Interpreter.loggerDebugL2,
				Level.INFO,
				"Running chunk {0} at offset {1}.",
				name(),
				offset)
		}
	}

	/**
	 * An enumeration of reasons why a chunk might be invalidated.
	 *
	 * @constructor
	 *
	 * @property countdownToNextOptimization
	 *   The number of invocations that must happen after this invalidation
	 *   before the code will be optimized into another chunk.
	 */
	enum class InvalidationReason constructor(
		val countdownToNextOptimization: Long)
	{
		/**
		 * The chunk is being invalidated because a method it depends on has
		 * changed.
		 */
		DEPENDENCY_CHANGED(200),

		/**
		 * The chunk is being invalidated due to it being evicted due to too
		 * many chunks being in existence.
		 */
		EVICTION(20000),

		/** The chunk is being invalidated to collect code coverage stats. */
		CODE_COVERAGE(200);

		/**
		 * [Statistic] for tracking the cost of invalidating chunks due to this
		 * reason.
		 */
		val statistic = Statistic(
			L2_OPTIMIZATION_TIME, "(invalidation from $name)")
	}

	/**
	 * Something that this `L2Chunk` depended on has changed. This must have
	 * been because it was optimized in a way that relied on some aspect of the
	 * available definitions (e.g., monomorphic inlining), so we need to
	 * invalidate the chunk now, so that an attempt to invoke it or return into
	 * it will be detected and converted into using the [unoptimizedChunk]. Also
	 * remove this chunk from the contingent set of each object on which it was
	 * depending.
	 *
	 * This can only happen when L2 execution is suspended, due to a method
	 * changing (TODO`MvG` - we'll have to consider dependent nearly-constant
	 * variables changing at some point).  The [invalidationLock] must be
	 * acquired by the caller to ensure safe manipulation of the dependency
	 * information.
	 *
	 * Note that all we do here is clear the valid flag and update the
	 * dependency information.  It's up to any re-entry points within this
	 * optimized code to determine that invalidation has happened,
	 * using the default chunk.
	 *
	 * @param reason
	 *   The [InvalidationReason] that indicates why this invalidation is
	 *   happening.
	 */
	fun invalidate(reason: InvalidationReason)
	{
		val before = AvailRuntimeSupport.captureNanos()
		assert(invalidationLock.isHeldByCurrentThread)
		AvailRuntime.currentRuntime().assertInSafePoint()
		assert(this !== unoptimizedChunk)
		isValid = false
		val contingents: A_Set = contingentValues.makeImmutable()
		contingentValues = emptySet
		for (value in contingents)
		{
			value.removeDependentChunk(this)
		}
		code?.setStartingChunkAndReoptimizationCountdown(
			unoptimizedChunk, reason.countdownToNextOptimization)
		Generation.removeInvalidatedChunk(this)
		val after = AvailRuntimeSupport.captureNanos()
		// Use interpreter #0, since the invalidationLock prevents concurrent
		// updates.
		reason.statistic.record(after - before, 0)
	}

	/**
	 * Dump the chunk to disk for debugging. This is expected to be called
	 * directly from the debugger, and should result in the production of three
	 * files: `JVMChunk_«uuid».l1`, `JVMChunk_«uuid».l2`, and
	 * `JVMChunk_«uuid».class`. This momentarily sets the
	 * [JVMTranslator.debugJVM] flag to `true`, but restores it to its original
	 * value on return.
	 *
	 * @return
	 *   The base name, i.e., `JVMChunk_«uuid»`, to allow location of the
	 *   generated files.
	 */
	@Suppress("unused")
	abstract fun dumpChunk(): String

	companion object
	{
		/**
		 * Answer a descriptive (non-unique) name for the specified
		 * [function][A_RawFunction].
		 *
		 * @param code
		 *   An arbitrary function, or `null` for the default `L2Chunk`.
		 * @return
		 *   The effective name of the function.
		 */
		fun name(code: A_RawFunction?): String =
			code?.methodName?.asNativeString() ?: "«default»"

		/**
		 * Each time an [A_RawFunction] is found to be the running code for some
		 * interpreter during periodic polling, atomically decrease its
		 * countdown by this amount, avoiding going below one (`1`).
		 *
		 * This temporal signal should be more effective at deciding what to
		 * optimize than just counting the number of times the code is called.
		 */
		const val decrementForPolledActiveCode: Long = 1000

		/**
		 * The [lock][ReentrantLock] that protects invalidation of chunks due to
		 * [method][MethodDescriptor] changes from interfering with each other.
		 * The alternative to a global lock seems to imply deadlock conditions.
		 */
		val invalidationLock = ReentrantLock()
	}
}
